hydroflow_lang/src/graph/ops/persist.rs

```diff
@@ -0,0 +1,155 @@
+///
+/// > Arguments: two Rust closures. The first generates an initial value per group. The second
+/// itself takes two arguments: an 'accumulator', and an element. The second closure returns the
+/// value that the accumulator should have for the next iteration.
+///
+/// A special case of `fold`, in the spirit of SQL's GROUP BY and aggregation constructs. The input
+/// is partitioned into groups by the first field, and for each group the values in the second
+/// field are accumulated via the closures in the arguments.
+///
+/// > Note: The closures have access to the [`context` object](surface_flows.md#the-context-object).
+///
+/// `group_by` can also be provided with one generic lifetime persistence argument, either
+/// `'tick` or `'static`, to specify how data persists. With `'tick`, values will only be collected
+/// within the same tick. With `'static`, values will be remembered across ticks and will be
+/// aggregated with pairs arriving in later ticks. When not explicitly specified persistence
+/// defaults to `'static`.
+///
+/// `group_by` can also be provided with two type arguments, the key and value type. This is
+/// required when using `'static` persistence if the compiler cannot infer the types.
+///
+/// ```hydroflow
+/// source_iter([("toy", 1), ("toy", 2), ("shoe", 11), ("shoe", 35), ("haberdashery", 7)])
+///     -> group_by(|| 0, |old: &mut u32, val: u32| *old += val)
+///     -> for_each(|(k, v)| println!("Total for group {} is {}", k, v));
+/// ```
+///
+/// Example using `'tick` persistence:
+/// ```rustbook
+/// let (input_send, input_recv) = hydroflow::util::unbounded_channel::<(&str, &str)>();
+/// let mut flow = hydroflow::hydroflow_syntax! {
+///     source_stream(input_recv)
+///         -> group_by::<'tick, &str, String>(String::new, |old: &mut _, val| {
+///             *old += val;
+///             *old += ", ";
+///         })
+///         -> for_each(|(k, v)| println!("({:?}, {:?})", k, v));
+/// };
+///
+/// input_send.send(("hello", "oakland")).unwrap();
+/// input_send.send(("hello", "berkeley")).unwrap();
+/// input_send.send(("hello", "san francisco")).unwrap();
+/// flow.run_available();
+/// // ("hello", "oakland, berkeley, san francisco, ")
+///
+/// input_send.send(("hello", "palo alto")).unwrap();
+/// flow.run_available();
+/// // ("hello", "palo alto, ")
+/// ```
+#[hydroflow_internalmacro::operator_docgen]
+pub const PERSIST: OperatorConstraints = OperatorConstraints {
+    name: "persist",
+    hard_range_inn: RANGE_1,
+    soft_range_inn: RANGE_1,
+    hard_range_out: RANGE_1,
+    soft_range_out: RANGE_1,
+    num_args: 0,
+    persistence_args: &(0..=1),
+    type_args: RANGE_0,
+    is_external_input: false,
+    ports_inn: None,
+    ports_out: None,
+    properties: FlowProperties {
+        // TODO(mingwei): Are these right?
+        deterministic: FlowPropertyVal::Preserve,
+        monotonic: FlowPropertyVal::Yes,
+        inconsistency_tainted: false,
```

Comment:
@jhellerstein need to take a look at this

Reply:
Concur.